tools/squidclient/squidclient.cc

```diff
@@ -425,18 +425,16 @@ main(int argc, char *argv[])
     }
 #endif
     /* Build the HTTP request */
+    const char *pathPassword = nullptr;
     if (strncmp(url, "mgr:", 4) == 0) {
-        char *t = xstrdup(url + 4);
-        const char *at = nullptr;
-        if (!strrchr(t, '@')) { // ignore any -w password if @ is explicit already.
-            at = ProxyAuthorization.password;
-        }
-        // embed the -w proxy password into old-style cachemgr URLs
-        if (at)
-            snprintf(url, sizeof(url), "cache_object://%s/%s@%s", Transport::Config.hostname, t, at);
-        else
-            snprintf(url, sizeof(url), "cache_object://%s/%s", Transport::Config.hostname, t);
-        xfree(t);
+        const auto path = url + 4;
+        const auto at = strrchr(path, '@'); // will ignore any -w password if @ is explicit already
+        int pathLen = at ? (at - path) : strlen(path);
+        if (at && *(at+1))
+            pathPassword = xstrdup(at+1);
+        const auto pathCopy = xstrdup(path);
+        snprintf(url, sizeof(url), "http://%s:%hu/squid-internal-mgr/%.*s", Transport::Config.hostname, Transport::Config.port, pathLen, pathCopy);
+        xfree(pathCopy);
```

Comment:
I see what you are doing with pathLen and pathCopy, but can we simplify along the following lines?

```suggestion
        char *t = xstrdup(url + 4);
        // XXX: Bail on snprintf() failures
        snprintf(url, sizeof(url), "http://%s:%hu/squid-internal-mgr/%s", Transport::Config.hostname, Transport::Config.port, t);
        if (const auto at = strrchr(url, '@')) {
            *at = 0; // send password in Proxy-Authorization header, not URL
            pathPassword = at + 1; // embedded @password overwrites -w password further below
        }
        xfree(t);
```

... and either remove xfree(pathPassword) at the bottom OR dupe that string in the above code.

I do not think we should ignore empty embedded passwords, so I did not add a check for `*pathPassword`. AFAICT, the official code did not check either.

Reply:
Done.